milli/src/search/criteria/exactness.rs

```diff
@@ -360,3 +329,452 @@ impl ExactQueryPart {
+///
+/// Therefore the time complexity is O(N^3 * M).
+fn create_non_disjoint_combinations(bitmaps: Vec<RoaringBitmap>) -> Vec<RoaringBitmap> {
+    let nbr_parts = bitmaps.len();
+    if nbr_parts == 1 {
+        let flattened_base_level = MultiOps::union(bitmaps.into_iter());
+        return vec![flattened_base_level];
+    }
+    let mut flattened_levels = vec![];
+    let mut last_level: BTreeMap<usize, RoaringBitmap> =
+        bitmaps.clone().into_iter().enumerate().collect();
+
+    for _ in 2..=nbr_parts {
+        let mut new_level = BTreeMap::new();
+        for (last_part_index, base_combination) in last_level.iter() {
+            #[allow(clippy::needless_range_loop)]
+            for new_last_part_index in last_part_index + 1..nbr_parts {
+                let new_combination = base_combination & &bitmaps[new_last_part_index];
+                if !new_combination.is_empty() {
+                    match new_level.entry(new_last_part_index) {
+                        Entry::Occupied(mut b) => {
+                            *b.get_mut() |= new_combination;
+                        }
+                        Entry::Vacant(entry) => {
+                            entry.insert(new_combination);
+                        }
+                    }
+                }
+            }
+        }
+        // Now flatten the last level to save memory
+        let flattened_last_level = MultiOps::union(last_level.values());
+        flattened_levels.push(flattened_last_level);
+        last_level = new_level;
+    }
+    // Flatten the last level
+    let flattened_last_level = MultiOps::union(last_level.values());
+    flattened_levels.push(flattened_last_level);
+    flattened_levels
+}
+
+/// Given a list of bitmaps `b0,b1,...,bn` , compute the list of bitmaps `X0,X1,...,Xn`
+/// such that `Xi` contains all the elements that are contained in **exactly** `i+1` bitmaps among `b0,b1,...,bn`.
+///
+/// The returned vector is guaranteed to be of length `n`. It is equal to `vec![X0, X1, ..., Xn]`.
+///
+/// ## Implementation
+/// 1. We first create `Y0,Y1,...Yn` such that `Yi` contains all the elements that are contained in
+/// **at least** `i+1` bitmaps among `b0,b1,...,bn`. This is done using `create_non_disjoint_combinations`.
+///
+/// 2. We create a set of "forbidden" elements, `Fn`, which is initialised to the empty set.
+///
+/// 3. We compute:
+///     - `Xn = Yn - Fn`
+///     - `Fn-1 = Fn | Xn`
+fn create_disjoint_combinations(parts_candidates_array: Vec<RoaringBitmap>) -> Vec<RoaringBitmap> {
+    let non_disjoint_combinations = create_non_disjoint_combinations(parts_candidates_array);
+
+    let mut disjoint_combinations = vec![];
+    let mut forbidden = RoaringBitmap::new();
+    for mut combination in non_disjoint_combinations.into_iter().rev() {
+        combination -= &forbidden;
+        forbidden |= &combination;
+        disjoint_combinations.push(combination)
+    }
+    disjoint_combinations.reverse();
+    disjoint_combinations
```

Comment:
I think we don't have to store the forbidden ids in a union of RoaringBitmap because `Yn+1` is always a subset of `Yn` making `Yn+1 | (Yn - Yn+1) == Yn`. Am I wrong? 🤔 

```suggestion
    let mut disjoint_combinations = vec![];
    let mut combinations = non_disjoint_combinations.into_iter().peekable();
    while let Some(mut combination) = combinations.next()  {
        if let Some(forbidden) = combinations.peek() {
            combination -= forbidden;
        }
        disjoint_combinations.push(combination)
    }
    
    disjoint_combinations
```

Reply:
I think you're right!